rosbag2_compression/src/rosbag2_compression/zstd_compressor.cpp

```diff
@@ -0,0 +1,120 @@
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+#include <chrono>
+#include <memory>
+#include <string>
+
+#include "logging.hpp"
+#include "rosbag2_compression/zstd_compressor.hpp"
+
+namespace
+{
+// Increasing the compression level will:
+//   - Increase the time taken to compress
+//   - Decrease the size of the compressed data
+// Setting to zero uses Zstd's default value of 3.
+constexpr const int DEFAULT_ZSTD_COMPRESSION_LEVEL = 1;
+
+std::unique_ptr<char[]> get_input_buffer(
+  const std::string & uri,
+  size_t & decompressed_buffer_length)
+{
+  try {
+    std::ifstream infile{uri};
+    infile.exceptions(std::ifstream::failbit);
+    // Get size and allocate
+    infile.seekg(0, std::ios::end);
+    decompressed_buffer_length = infile.tellg();
+    auto decompressed_buffer = std::make_unique<char[]>(decompressed_buffer_length);
+    // Go back and read in contents
+    infile.seekg(0 /* off */, std::ios::beg);
+    infile.read(decompressed_buffer.get(), decompressed_buffer_length);
+    return decompressed_buffer;
+  } catch (std::ios_base::failure & fail) {
+    ROSBAG2_COMPRESSION_LOG_ERROR_STREAM(
+      "Caught IO stream exception while reading file: " << fail.what());
+    throw fail;
+  }
+}
+
+void check_compression_result(const size_t compression_result)
+{
+  if (ZSTD_isError(compression_result)) {
+    std::stringstream error;
+    error << "ZSTD compression error: " << ZSTD_getErrorName(compression_result);
+    throw std::runtime_error(error.str());
+  }
+  ROSBAG2_COMPRESSION_LOG_DEBUG("ZSTD compressed file.");
+}
+
+void print_compression_statistics(
+  std::chrono::high_resolution_clock::time_point start,
+  std::chrono::high_resolution_clock::time_point end,
+  size_t decompressed_size, size_t compressed_size)
+{
+  const auto duration = std::chrono::duration_cast<std::chrono::microseconds>(end - start);
+  const auto compression_ratio =
+    static_cast<double>(decompressed_size) / static_cast<double>(compressed_size);
+  ROSBAG2_COMPRESSION_LOG_DEBUG_STREAM(
+    "Compression statistics:\n" <<
+      "Time: " << duration.count() << " microseconds" <<
+      "Compression Ratio: " << compression_ratio
+  );
+}
+}  // namespace
+
+namespace rosbag2_compression
+{
+
+std::string ZstdCompressor::compress_uri(const std::string & uri)
+{
+  const auto start = std::chrono::high_resolution_clock::now();
+  const auto compressed_uri = uri + "." + get_compression_identifier();
+  try {
+    size_t decompressed_buffer_length{0};
+    const auto decompressed_buffer = get_input_buffer(uri, decompressed_buffer_length);
+    // Allocate based on compression bound and compress
+    const size_t compressed_buffer_length = ZSTD_compressBound(decompressed_buffer_length);
+    auto compressed_buffer = std::make_unique<char[]>(compressed_buffer_length);
```

Comment:
I actually have two problem here. Why using a C-Style array? why not using a vector? Why do we need a unique_ptr? Just for not cleaning up in case of error?
Also, I still think that `char` is not the right type for a general buffer - I think the preferred type is `uint8_t` as it's clearly defined on all platforms.

Reply:
> Why do we need a unique_ptr?

1. To handle scope and cleaning up in case of an error.
2.  zstd_compress needs a raw pointer to the data to compress/output decompressed buffer.

Regarding `uint8_t`, working on it...

Reply:
+1 on the C array, this should not be necessary.

Reply:
an `std::vector<std::uint8_t>` should cover both of your concerns. It does clean up its internal memory and with `std::vector<T>::data()` you gain access to its `T*` instance.

Reply:
I've changed my mind on this one, `get_input_buffer` should have the signature:
```c++
std::vector<uint8_t> get_input_buffer(const std::string & uri)
```
since `std::vector` includes the size of the data buffer.

There's no real difference for `compressed_buffer` since it exists in the same scope as `compressed_buffer_length`.